third_party/move/move-compiler-v2/src/pipeline/livevar_analysis_processor.rs

```diff
@@ -41,9 +93,378 @@ impl FunctionTargetProcessor for LiveVarAnalysisProcessor {
+        }
+    }
+}
+
+// =================================================================================================
+// Bytecode Transformation
+
+/// State for copy inference transformation.
+struct CopyTransformation<'a> {
+    fun_env: &'a FunctionEnv<'a>,
+    data: FunctionData,
+}
+
+impl<'a> CopyTransformation<'a> {
+    /// Runs copy inference transformation. This transformation inserts implicit copies. It also
+    /// checks correctness of copies, whether explicit or implicit.
+    fn transform(&mut self, alive: &LiveVarAnnotation) {
+        let code = std::mem::take(&mut self.data.code);
+        for (i, bc) in code.into_iter().enumerate() {
+            self.transform_bytecode(
+                alive
+                    .get_live_var_info_at(i as CodeOffset)
+                    .expect("live var info"),
+                bc,
+            )
+        }
+    }
+
+    /// Transforms a bytecode. This handles `Assign` and `Call` instructions.
+    /// For the former, it infers the `AssignKind` (copy or move) and for the later,
+    /// it implicitly copies arguments if needed. Implicit copy is needed
+    /// if a non-primitive value is used after the given program point.
+    fn transform_bytecode(&mut self, alive: &LiveVarInfoAtCodeOffset, bc: Bytecode) {
+        use Bytecode::*;
+        match bc {
+            Assign(id, dst, src, kind) => match kind {
+                AssignKind::Inferred => {
+                    if self.check_implicit_copy(alive, id, false, src) {
+                        self.data.code.push(Assign(id, dst, src, AssignKind::Copy))
+                    } else {
+                        self.data.code.push(Assign(id, dst, src, AssignKind::Move))
+                    }
+                },
+                AssignKind::Copy | AssignKind::Store => {
+                    self.check_explicit_copy(id, src);
+                    self.data.code.push(Assign(id, dst, src, AssignKind::Copy))
+                },
+                AssignKind::Move => {
+                    self.check_explicit_move(alive, id, src);
+                    self.data.code.push(Assign(id, dst, src, AssignKind::Move))
+                },
+            },
+            Call(_, _, Operation::BorrowLoc, _, _) => {
+                // Borrow does not consume its operand and need no copy
+                self.data.code.push(bc)
+            },
+            Call(id, dsts, oper, srcs, ai) => {
+                let srcs = self.copy_arg_if_needed(alive, id, srcs);
+                self.data.code.push(Call(id, dsts, oper, srcs, ai))
+            },
+            _ => self.data.code.push(bc),
+        }
+    }
+
+    /// Walks over the argument list and inserts copies if needed.
+    fn copy_arg_if_needed(
+        &mut self,
+        alive: &LiveVarInfoAtCodeOffset,
+        id: AttrId,
+        srcs: Vec<TempIndex>,
+    ) -> Vec<TempIndex> {
+        use Bytecode::*;
+        let mut new_srcs = vec![];
+        for (i, src) in srcs.iter().enumerate() {
```

Comment:
into_iter() here.  This function is already consuming srcs, can get rid of some copies/clones below.

Reply:
Oh, yeah.